drivers/ina2xx/ina2xx.c

```diff
@@ -0,0 +1,190 @@
+ *
+ * @}
+ */
+
+#include <errno.h>
+#include <stdint.h>
+
+#include "ina2xx.h"
+#include "ina2xx_defines.h"
+#include "periph/i2c.h"
+#include "byteorder.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+/** @brief Read one 16 bit register from a INA2XX device and swap byte order, if necessary. */
+static int ina2xx_read_reg(const ina2xx_t *dev, uint8_t reg, uint16_t *out)
+{
+    union {
+        uint8_t c[2];
+        uint16_t u16;
+    } tmp = { .u16 = 0 };
+    int status = 0;
+
+    status = i2c_read_regs(dev->params.i2c, dev->params.addr, reg, &tmp.c[0],
+                           2, 0);
+
+    if (status < 0) {
+        return status;
+    }
+
+    *out = ntohs(tmp.u16);
+    return 0;
+}
+
+/** @brief Write one 16 bit register to a INA2XX device and swap byte order, if necessary. */
+static int ina2xx_write_reg(const ina2xx_t *dev, uint8_t reg, uint16_t in)
+{
+    union {
+        uint8_t c[2];
+        uint16_t u16;
+    } tmp = { .u16 = 0 };
+    int status = 0;
+
+    tmp.u16 = htons(in);
+
+    status = i2c_write_regs(dev->params.i2c, dev->params.addr, reg, &tmp.c[0],
+                            2, 0);
+
+    if (status < 0) {
+        return status;
+    }
+
+    return 0;
+}
+
+
+int ina2xx_init(ina2xx_t *dev, const ina2xx_params_t *params)
+{
+    uint16_t config;
+    int status;
+    if (!dev || !params) {
+        return -EINVAL;
+    }
+
+    dev->params = *params;
+    /* Reset device */
+    status = ina2xx_write_reg(dev, INA2XX_REG_CONFIGURATION, INA2XX_RESET);
+    if (status < 0) {
+        return status;
+    }
+
+    /* Check if default config is preset after reset */
+    status = ina2xx_read_reg(dev, INA2XX_REG_CONFIGURATION, &config);
+    if (status < 0) {
+        return status;
+    }
+
+    if (config != INA2XX_DEFCONFIG) {
+        DEBUG("[ina2xx]: Reset did't restore default config. Wrong device?\n");
+        return -ENODEV;
+    }
+
+    status = ina2xx_write_reg(dev, INA2XX_REG_CONFIGURATION, params->config);
+    if (status < 0) {
+        return status;
+    }
+
+    /* Set calibration so that current register resolution is 0.01 mA */
+    uint32_t calib = 4096000 / params->rshunt_mohm;
+    /* Device by 2^i_range to reduce the resolution by factor 2^i_range */
+    calib >>= params->i_range;
+
+    if (calib > UINT16_MAX) {
+        return -ERANGE;
+    }
+
+    return ina2xx_write_reg(dev, INA2XX_REG_CALIBRATION, (uint16_t)calib);
+}
+
+int ina2xx_read_shunt(const ina2xx_t *dev, int16_t *voltage)
+{
+    return ina2xx_read_reg(dev, INA2XX_REG_SHUNT_VOLTAGE, (uint16_t *)voltage);
+}
+
+int ina2xx_read_bus(const ina2xx_t *dev, uint16_t *voltage)
+{
+    uint16_t tmp;
+    int status = ina2xx_read_reg(dev, INA2XX_REG_BUS_VOLTAGE, &tmp);
+    if (status < 0) {
+        return status;
+    }
+
+    /* The voltage given by bits 15-3 in steps of 4 mV
+     * ==> shift 3 to the right and multiply by 4 to get voltage in mV
+     */
+    *voltage = (tmp >> 3) * 4;
```

Comment:
For a moment I thought the compiler could optimize it wrong, but probably not.

```suggestion
    *voltage = (tmp & ~0x7) >> 1;
```

will still generate [slightly smaller](https://godbolt.org/z/Bh9vXg) code, at least on AVR.

Reply:
don